javaslang/src/main/java/javaslang/collection/TreeMap.java

```diff
@@ -477,6 +505,98 @@ public boolean isEmpty() {
         return TreeSet.ofAll(keyComparator(), iterator().map(Tuple2::_1));
     }
 
+    @Override
+    public Iterator<Tuple2<K, V>> descendingIterator() {
+        return entries.descendingIterator();
+    }
+
+    @Override
+    public SortedSet<K> descendingKeySet() {
```

Comment:
In TreeMap the return type needs to be TreeMap, i.e.

```java
    ...
    TreeMap<K, V> descendingMap();
    ...
    TreeMap<K, V> headMap(K toKey);
    TreeMap<K, V> headMap(K toKey, boolean inclusive);
    ...
    TreeMap<K, V> tailMap(K fromKey);
    TreeMap<K, V> tailMap(K fromKey, boolean inclusive);
    ...
    TreeMap<K, V> subMap(K fromKey, boolean fromInclusive, K toKey, boolean toInclusive);
    TreeMap<K, V> subMap(K fromKey, K toKey);
    TreeMap<K, V> subMap(Option<K> fromKey, boolean fromInclusive, Option<K> toKey, boolean toInclusive);
    ...
```

Please note that `descendingKeySet()` and `keySet()` for example return `SortedSet` instead of a concrete implementation type. That's ok. The current 'rule'  is that a type only adjusts the self-return types (like TreeMap).

While reviewing this PR I've questioned the return type adjustment _in implementation types_. Of cause interfaces need to adjust them. But I see no benefit in adjusting the return type of methods within implementing classes.

Scala does adjust return types (e.g. in `scala.collection.immutable.TreeMap`), Java does not (e.g. in `java.util.TreeMap`).

I think it does not hurt to adjust return types, so we will go on like that in Javaslang.